### ISSUE TITLE
Clarify behavior of float division/remainder by 0

### ISSUE THREAD

**Issue by teoxoy** [gpuweb/gpuweb#2798]
Currently the spec mentions that division/modulo by 0 on integers results in 0 (i.e. `x / 0 = 0` and `x % 0 = 0`).

![image](https://user-images.githubusercontent.com/28601907/165554477-22c659ec-3794-442d-aaeb-6df453a54bd9.png)

I think it's important to clarify this for floats as well since at least SPIR-V's `OpFRem` (haven't looked at the behavior of other backends yet) seems to have undefined behavior if the 2nd operand is 0.

![image](https://user-images.githubusercontent.com/28601907/165555909-8bb081bc-ef5a-4b34-b5fb-d8233026d1e6.png)

**Comment by m-schuetz** [comment]
In glsl, 1.0 / 0.0 gives Infinity and -1.0 / 0.0 gives -Infinity which is good to have since there is no predefined infinity constant. Would be good to either have an Infinity constant or allow computing infinity via 1.0 / 0.0 in WebGPU.

**Comment by jzm-intel** [comment]
I thought the spec said `x / 0 = x`.
> If T is a signed integral type, the scalar case, evaluates to:
>   * e1, when e2 is zero.

**Comment by jzm-intel** [comment]
I see, `e1 / e2` has no clear definition in this table.
However, for `e1 % e2` we do have the result equals to `e1 - e2 * trunc(e1 / e2)`. Let's say for floating point `x`, `x / 0.0 == x`, then we will have `x % 0.0 == x - 0.0 * trunc(x / 0.0) == x - 0.0 * trunc(x) == x` rather than `x % 0.0 == 0.0`. 
I am not sure, maybe the bahavior of `x / 0.0` for a floating point `x` is left undefined intentionally and its behavior relays on backends?

**Comment by teoxoy** [comment]
> I am not sure, maybe the bahavior of `x / 0.0` for a floating point `x` is left undefined intentionally and its behavior relays on backends?

IIRC all backends have undefined behavior for division/remainder by 0 which I think it's a portability issue for WGSL.

**Comment by jzm-intel** [comment]
I understand, thanks

**Comment by teoxoy** [comment]
Something worth pointing out is that if there is no consensus on this, the decision might impact #2765 (due to some of the built-in functions being defined in terms of div/rem).

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2022-05-10-minutes#clarify-behavior-of-float-divisionremainder-by-0--issue-2798">WGSL meeting minutes 2022-05-10</a></summary>



* AB: Float div by zero is well defined. We don’t really support infinities so ends up being undefined value. Referencing spirv it’s undefined value vs scary undefined behaviour. Dont’ think we can do better then undefined value.
* KG: Arbitrary value? Afraid of calling undefined values, if you branch based on undefined value you get undefined behaviour. Any word that means undefined without saying undefined we’re in a better position. Floating arbitrary value instead
* AB: **Separate issue. Should lump them all together and figure out what we want to say about undefined value. If text already in spec that’s already undefined in other languages (extracting from composite). We should make decision on whole.**
* KG: Sounds great
* MM: Performance vs portability trade off?
* KG: Infinities aren’t always extant on the platform. Ideally the implementation gives you something. Decides i don’t do infinity so gives max value or 0 or maybe 3. Ideally, restriction trying to draw this isnt’ undefined value over branches causing UB. Would it be less scary if saying this is what we’re calling an undefined value which can not cause ub? \
MM: Less scary. Less worried about portability. Should investigate div by zero polyfil for integral values to see cost
* DN: Division by epsilon is also overflow. Problems occurring at 0 also occur close to zero. So no useful polyfil without being conservitive and making div not div anymore
* KG: If investing if polyfil will work, need idea of what polyfill would be. What would you want it to be
* MM: Don’t know …
* KG: Can I task you with that?
* MM: Probability low. Not going to stand in way, just stating values.
* KG: Leaves with something where we don’t technically need to do anything except for
* AB:  &lt;chat message>
    * Alan Baker8:13 PM
    * what about (roughly):
    * x / y = round(y) == 0 ? 0 : x/y
    * 
    * If you preround the value potentially, think it might cover other cases. Woudln’t be happy about making floating point slow but potentially something like this could work. Don’t know how many devices we’d need to test on.
* KG: Would polyfill and normalize just this case, but infinity on load would still get undefined values. Having our cake and eating it to in that maybe we have infinity
* AB: If have infinity the frem case maybe bad but hte ?? case i don’t know. The ?? If  supporting infinities don’t know what you’d get out of that
* KG: If you put x/y in polyfil and compared with infinity from load/store then not equal unless platform doesn’t have infinities and normalizes into zeros.
* AB: Just picked 0 because we picked for integer division
* KG: Part of exercise is figuring out where boundaries of portability vs allowing infinities is. 1/0 has well defined value if platform has infinity. Let’s think more outside time box.


</details>

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2022-05-31-minutes#clarify-behavior-of-float-divisionremainder-by-0--issue-2798">WGSL meeting minutes 2022-05-31</a></summary>



* KG: Planned on discussion of what it should be
* DN: Unlike int division, ieee has defined 1/0. That’s why different. Think this is a sub-answer for 2777 because the problem is some implementations don’t have inf or nan. If you have calculation which has one, what do we say? Want to push on stack until we solve 2777
* KG: Agree it should be generic. Basically just defined as ieee and additionally if yo umake inf or nan and the impl doesn’t support then extra step kicks in. SPec wise, you got inf, but system doesn’t do so you get something else. If we really wanted to support (like div/0) could spec it a particular way and that’s where we would say it.
* MM: Think that's reasonable. Biggest concern is branch on this value is _not_ UB.
* KG: Concern being requirement?
* MM: Yes. Whatever we do, branching on value should not be UB. Don’t know how to implementat in SPIR-V.
* KG: Tolerable if unspecified value and impl can have different ones?
* MM: Yes, ideally not but don’t see way around in this case. Trade off is if we want to spec how they work then we’d be leaving a lot of android devices out. Past that we want to make sure that (e.g. branch on if/else then one of the blocks, not both, not neither is taken).
* KG: Ties into what nouns we use for undefined values and behaviours. Different pr and issue. **That’s waiting on PR from KG**. Will finish doing that and we’ll have better nouns. Sounds like we’re on the same page. Satisfied for now. 


</details>

**Comment by djg** [comment]
> In glsl, 1.0 / 0.0 gives Infinity and -1.0 / 0.0 gives -Infinity which is good to have since there is no predefined infinity constant. Would be good to either have an Infinity constant or allow computing infinity via 1.0 / 0.0 in WebGPU.

The C standard library[1] defines `HUGE_VAL` as the result of computations that produce errors. This value is implementation-defined and maps to `inf` on a number of compilers.

In `fast-math` environments, compilers can assume that `inf` is never produced and optimize accordingly, potentially removing `isinf()` checks, so I suggest introducing a `HUGE_VAL` equivalent defined as the maximum representable floating point number to be returned by `x / 0.0`, `ldexp()`, et al.

[1] https://en.cppreference.com/w/c/numeric/math/ldexp

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2022-10-18-minutes#clarify-behavior-of-float-divisionremainder-by-0--issue-2798">WGSL 2022-10-18 Minutes</a></summary>

* DG: Looked at ldexp in stdlib, defined to return HUGE_VAL for error conditions. Talking internally, we think we need something like this for a frozen value, so that we can test about it.
* AB: There’s PR#3508, which I think does describe a solution like that for “indeterminate value”.
* DG: Do we expect to define this val?
* KG: Implementation defined.
* JB: We want to ensure that e.g. Inf is still valid if supported.
* **Waiting on DN to merge #3508.**

</details>
